application/controllers/QuestionAdministrationController.php

```diff
@@ -2785,50 +2794,73 @@ function ($collector, $oDefaultValue) use ($oQuestion) {
     }
 
     /**
-     * Save subquestion.
+     * Delete sub questions that no longer appear in the post request.
      * Used when survey is *not* activated.
      *
      * @param Question $question
      * @param array $subquestionsArray Data from request.
      * @return void
      * @throws CHttpException
      */
-    private function storeSubquestions($question, $subquestionsArray)
+    private function deleteMissingSubquestions($question, $subquestionsArray)
+    {
+        foreach ($question->subquestions as $subquestion) {
+            if (!isset($subquestionsArray[$subquestion->qid])) {
+                if (!$subquestion->delete()) {
+                    throw (new LSUserException(500, gT("Could not delete subquestion")))
+                        ->setDetailedErrorsFromModel($subquestion);
+                }
+            }
+        }
+    }
+
+   /**
+     * Add new subquestions.
+     * Used when survey is *not* activated.
+     *
+     * @param Question $question
+     * @param array $subquestionsArray Data from request.
+     * @return void
+     * @throws CHttpException
+    */
+    private function addNewSubquestions($question, $subquestionsArray)
     {
         $questionOrder = 0;
         $errorQuestions = [];
         foreach ($subquestionsArray as $subquestionId => $subquestionArray) {
-            foreach ($subquestionArray as $scaleId => $data) {
-                $subquestion = new Question();
-                $subquestion->sid        = $question->sid;
-                $subquestion->gid        = $question->gid;
-                $subquestion->parent_qid = $question->qid;
-                $subquestion->question_order = $questionOrder;
-                $questionOrder++;
-                if (!isset($data['code'])) {
-                    throw new CHttpException(
-                        500,
-                        'Internal error: Missing mandatory field code for question: ' . json_encode($data)
-                    );
-                }
-                $subquestion->title      = $data['code'];
-                if ($scaleId === 0) {
-                    $subquestion->relevance  = $data['relevance'];
-                }
-                $subquestion->scale_id   = $scaleId;
-                if (!$subquestion->save()) {
-                    array_push($errorQuestions, $subquestion);
-                    continue;
-                }
-                $subquestion->refresh();
-                foreach ($data['subquestionl10n'] as $lang => $questionText) {
-                    $l10n = new QuestionL10n();
-                    $l10n->qid = $subquestion->qid;
-                    $l10n->language = $lang;
-                    $l10n->question = $questionText;
-                    if (!$l10n->save()) {
-                        throw (new LSUserException(500, gT("Could not save subquestion")))
-                            ->setDetailedErrorsFromModel($l10n);
+            if (substr($subquestionId, 0, 3) === 'new') {
```

Comment:
This is very personal, but I like better this form of `if` when not having an else and inside a loop:
```
foreach() {
   if(!xxx) {
     continue;
   }
}
```
Flatter code.

Reply:
Also, isn't a supposition all in `subquestionsArray` will be new?
I feel some behaviour should be moved out of this function as to have a single objetive.
Right now it has 2:
- distinguish what to add new and what not.
- add new.

Reply:
:+1: here : really more clear !

Reply:
> Also, isn't a supposition all in `subquestionsArray` will be new?

No : it's the reason of the issue : loosing default value on existing subquestion

Reply:
Still, I think this applies.

> I feel some behaviour should be moved out of this function as to have a single objetive.
> Right now it has 2:
> - distinguish what to add new and what not.
> - add new.

Reply:
> Still, I think this applies.
> 
> > I feel some behaviour should be moved out of this function as to have a single objetive.
> > Right now it has 2:
> > 
> > * distinguish what to add new and what not.
> > * add new.

Something like this : 

```
function addOrUpdateSubquestions($question, $subquestionsArray) {
    $questionOrder = 0;
    foreach ($subquestionsArray as $subquestionId => $subquestionArray ) {
        if (substr($subquestionId, 0, 3) === 'new') {
            $this->addNewSubquestion($question, $subquestionArray, $questionOrder);
        } else {
            $this->updateSubquestion($question, $subquestionId, $subquestionArray, $questionOrder);
        }
    }
}
```

?

Reply:
Yes... something like that.
addOrUpdateSubquestions could be in the controller while updateSubquestion and addNewSubquestion on the service helper.